### PR TITLE
Crimson pool quota

### PR DESCRIPTION
**Pull request by liu-chunmei**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/pg.cc [hunk 1]
```diff
@@ -681,6 +681,37 @@ PG::do_osd_ops_execute(
...
+      if (m.get_reqid().name.is_mds() ||   // FIXME: ignore MDS for now
+        m.has_flag(CEPH_OSD_FLAG_FULL_FORCE)) {
+        logger().info(" full, but proceeding due to FULL_FORCE or MDS");
+      } else if (m.has_flag(CEPH_OSD_FLAG_FULL_TRY)) {
+        // they tried, they failed.
+        logger().info(" full, replying to FULL_TRY op");
+        if (get_pool().info.has_flag(pg_pool_t::FLAG_FULL_QUOTA))
+          return interruptor::make_ready_future<OpsExecuter::rep_op_fut_tuple>(
+            seastar::now(),
+            OpsExecuter::osd_op_ierrorator::future<>
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:698]:
```
             OpsExecuter::osd_op_ierrorator::future<>(
               crimson::ct_error::edquot::make()));
```
and elsewhere

File: src/crimson/osd/pg.h [hunk 2]
```diff
@@ -797,6 +797,9 @@ struct PG::do_osd_ops_params_t {
   uint64_t get_features() const {
     return features;
   }
+  bool has_flag(uint32_t flag) const {
```
Comment (athanatos) [inline src/crimson/osd/pg.h:800]: Do we still need to pull this from the actual message?

Comment (liu-chunmei) [reply src/crimson/osd/pg.h:800]: do_osd_ops( ... do_osd_ops_params_t ..) seems only is called in internal client request, no error handle, no return message (void), so I don't think it need check pool quota.

Comment (athanatos) [reply src/crimson/osd/pg.h:800]: Add a comment:

```
  // Only used by InternalClientRequest, no op flags
```

Comment (liu-chunmei) [reply src/crimson/osd/pg.h:801]: done.

**Comment by liu-chunmei** [comment]
jenkins test ceph API

File: src/crimson/osd/pg.h [hunk 3]
```diff
@@ -797,6 +797,10 @@ struct PG::do_osd_ops_params_t {
   uint64_t get_features() const {
     return features;
   }
+  //Only used by InternalClientRequest, no op flags
```
Comment (athanatos) [inline src/crimson/osd/pg.h:800]: Add a space between the // and the Only.

**Review by athanatos — APPROVED** [review]
LGTM other than formatting nit

**Comment by ronen-fr** [comment]
@liu-chunmei - note the missing actual description in the PR comment

**Pull request merged**